src/MessageTrait.php

```diff
@@ -170,8 +178,11 @@ public function withHeader($header, $value)
             );
         }
 
+        $lowerHeader = strtolower($header);
+
         $new = clone $this;
-        $new->headers[$header] = $value;
+        $new->headerNames[$lowerHeader] = $header;
+        $new->headers[$lowerHeader] = $value;
```

Comment:
`getHeader('X-Foo')` and `getHeader('X-foo')` should return the same thing. `getHeaders()` should return the header names as they were registered. RFC 7230 indicates that header names are case insensitive; however, some clients and servers do not respect this, so we have indicated in PSR-7 at this point that case should be preserved internally, while allowing querying (`hasHeader()`, `getHeader()`, `withoutHeader()`) should be case insensitive.

Now, where things get tricky: if the same header is registered multiple times using different case. I think in that case, we would use whatever casing was first registered.